wrapper/src/main/java/software/amazon/jdbc/ConnectionPluginChainBuilder.java

```diff
@@ -0,0 +1,229 @@
+        }
+      };
+
+  protected static final String DEFAULT_PLUGINS = "auroraConnectionTracker,failover,efm";
+
+  private static class PluginFactoryInfo {
+    public Class<? extends ConnectionPluginFactory> factory;
+    public int weight;
+
+    public PluginFactoryInfo(final Class<? extends ConnectionPluginFactory> factory, final int weight) {
+      this.factory = factory;
+      this.weight = weight;
+    }
+  }
+
+  public List<ConnectionPlugin> getPlugins(
+      final PluginService pluginService,
+      final ConnectionProvider defaultConnProvider,
+      final PluginManagerService pluginManagerService,
+      final Properties props)
+      throws SQLException {
+
+    List<ConnectionPlugin> plugins;
+    List<Class<? extends ConnectionPluginFactory>> pluginFactories;
+
+    final String profileName = PropertyDefinition.PROFILE_NAME.getString(props);
+
+    if (profileName != null) {
+
+      if (!DriverConfigurationProfiles.contains(profileName)) {
+        throw new SQLException(
+            Messages.get(
+                "ConnectionPluginManager.configurationProfileNotFound",
+                new Object[] {profileName}));
+      }
+      pluginFactories = DriverConfigurationProfiles.getPluginFactories(profileName);
+
+    } else {
+
+      String pluginCodes = PropertyDefinition.PLUGINS.getString(props);
+
+      if (pluginCodes == null) {
+        pluginCodes = DEFAULT_PLUGINS;
+      }
+
+      final List<String> pluginCodeList = StringUtils.split(pluginCodes, ",", true);
+      pluginFactories = new ArrayList<>(pluginCodeList.size());
+
+      for (final String pluginCode : pluginCodeList) {
+        if (!pluginFactoriesByCode.containsKey(pluginCode)) {
+          throw new SQLException(
+              Messages.get(
+                  "ConnectionPluginManager.unknownPluginCode",
+                  new Object[] {pluginCode}));
+        }
+        pluginFactories.add(pluginFactoriesByCode.get(pluginCode));
+      }
+    }
+
+    if (!pluginFactories.isEmpty()) {
+
+      if (!PropertyDefinition.PRESERVE_PLUGINS_ORDER.getBoolean(props)) {
+        pluginFactories = this.sortPluginFactories(pluginFactories);
+
+        final List<Class<? extends ConnectionPluginFactory>> tempPluginFactories = pluginFactories;
+        LOGGER.finest(() ->
+            "Plugins order has been rearranged. The following order is in effect: "
+                + tempPluginFactories.stream()
+                  .map(Class::getSimpleName)
+                  .collect(Collectors.joining(", ")));
+      }
+
+      try {
+        final ConnectionPluginFactory[] factories =
+            WrapperUtils.loadClasses(
+                    pluginFactories,
+                    ConnectionPluginFactory.class,
+                    "ConnectionPluginManager.unableToLoadPlugin")
+                .toArray(new ConnectionPluginFactory[0]);
+
+        // make a chain of connection plugins
+
+        plugins = new ArrayList<>(factories.length + 1);
```

Comment:
What is the value of initialize the plugins arraylist here with a fixed size? Can we just initialize it on line 118 
`final List<ConnectionPlugin> plugins = new ArrayList<>();` and remove the if-else clause.

Reply:
Using `new ArrayList()` will initialize the list with 0 capacity. Every time we need to add a new element to the list will require the list to expand. Since `ArrayList` is based on internal array it means that an internal copy from one array to a new array will occur. By specifying an initial size we can avoid unnecessary overhead with copying.